services/blockstorage/service.go

```diff
@@ -74,19 +77,32 @@ func (s *service) CommitBlock(input *services.CommitBlockInput) (*services.Commi
 		return nil, err
 	}
 
-	s.updateLastCommittedBlockHeightAndTimestamp(txBlockHeader)
+	s.updateLastCommittedBlock(input.BlockPair)
 
 	s.reporting.Info("Committed a block", instrumentation.BlockHeight(txBlockHeader.BlockHeight()))
 
 	return nil, nil
 }
 
-func (s *service) updateLastCommittedBlockHeightAndTimestamp(txBlockHeader *protocol.TransactionsBlockHeader) {
+func (s *service) updateLastCommittedBlock(block *protocol.BlockPairContainer) {
 	s.lastBlockLock.Lock()
 	defer s.lastBlockLock.Unlock()
 
-	s.lastCommittedBlockHeight = txBlockHeader.BlockHeight()
-	s.lastCommittedBlockTimestamp = txBlockHeader.Timestamp()
+	s.lastCommittedBlock = block
+}
+
+func (s *service) lastCommittedBlockHeight() primitives.BlockHeight {
+	if s.lastCommittedBlock == nil {
+		return 0
+	}
+	return s.lastCommittedBlock.TransactionsBlock.Header.BlockHeight()
+}
+
+func (s *service) lastCommittedBlockTimestamp() primitives.TimestampNano {
+	if s.lastCommittedBlock == nil {
+		return 0
+	}
+	return s.lastCommittedBlock.TransactionsBlock.Header.Timestamp()
```

Comment:
It's scary to me that this code is not under mutex.. `s.lastCommittedBlock` might be updated after the `if` is ran but before the return runs. So the if result might not be relevant. Let's assume that `s.lastCommittedBlock` could be updated externally to nil, this would mean that the return will crash..

Reply:
not sure i see how read protect will assist here
its nil on initialization, this is just good practice, but its never really nil
and if its updated *after* the `if`, read protection will not assist with getting a more accurate answer (no interrupt control can actually, the `if` already happened)

i am going to do some research on go to be sure, but the read operation in line 105, as far as i know, will be executed atomically, the `timestamp()` call will not, but its data fetch will, and if we are going to assume that between 102->105 the lastCB will be zeroed out, then a lot more than our code is unsafe in that matter (every struct call need to be protected? and the protection is not only around the read in that case, its around the entire execution flow) and again, afaik there is no line of code that will do `s.lCB = nil`